### DISCUSSION TITLE
BUG_Failed to get accessToken, Network request failed

### DISCUSSION THREAD

**Discussion by kjakobczyk** [victorbalssa/abacus#41] [Bug] [answered]
**Describe the bug**
I'm trying to use Abacus with Firefly III server v.5.7.9 running on Docker via `docker-compose`.

**To Reproduce**
1. Create new OAuth token in Firefly III
2. Open Abacus for Firefly III
3. Fill the backend URL and Client ID in Abacus for Firefly III leaving client secret empty
4. Click on "Sign in" button
5. Log in to Firefly III
6. Click on "Authorize" in Firefly III
7. You see the error message saying that "Something went wrong" with description "failed to get accessToken, Network request failed.

**Expected behavior**
After clicking on  "Authorize" in Firefly III I'd expect the application to log me in and allow to insert information to my Firefly III instance through Abacus for Firefly.

**Screenshots**
![obraz](https://user-images.githubusercontent.com/44730139/190976036-2d9a464e-ce82-4fbf-9ed7-7123d59df37f.png)

**Smartphone (please complete the following information):**
 - Device: iPhone 13
 - OS: iOS 16
 - Firefly III v5.7.9 running in Docker behind Apache2 reverse proxy
 - Latest Abacus for Firefly III

**Comment by victorbalssa** [comment, marked as answer]
Do you have any errors after **step 6**?

What happen after step 6? safari close the current connexion and you get the red notification?
If that's the case, can you look at your Firefly III logs and display what errors you may have?

**Reply by kjakobczyk** [reply]
After step 6 the red error message from step 7 is being shown. I can ignore the message and click "Sign in" button once again, but this brings me back to step 4, although without the need to enter the username/password and the problem repeats the same way. Even if I repeat the process several times in a row the outcome is the same.

Regarding Firefly III logs: I have recreated the container with `APP_DEBUG=true` and `LOG_LEVEL=debug`. Looking at the `stdout` logs from `docker logs -f [container_id]` the only things I see after steps 4 and 6 are the following:

```
172.26.0.1 - - [19/Sep/2022:21:46:18 +0200] "GET /oauth/authorize?code_challenge=Xn4cNM-QlABhw0Fo4vSNQSfCBOhG5yA4zZMS7ZaKVck&code_challenge_method=S256&redirect_uri=abacusiosapp%3A%2F%2Fredirect&client_id=3&response_type=code&state=cQTlhQc9jp HTTP/1.1" 200 3672 "-" "Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.0 Mobile/15E148 Safari/604.1"
172.26.0.1 - - [19/Sep/2022:21:46:36 +0200] "POST /oauth/authorize HTTP/1.1" 302 2652 "-" "Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.0 Mobile/15E148 Safari/604.1"
```

**Reply by kjakobczyk** [reply]
OK, I've got an update!

When I omitted the Apache2 reverse proxy and connected directly to Firefly III docker instance I was able to sign in!
The only thing I **had** to do was to provide both `Client ID` and `Client secret`. Without `Client secret` it was not working.

Anyway, it seems that this is some kind of misconfiguration in my reverse proxy setup. I will need to review that. Thank you for your engagement!

**Reply by victorbalssa** [reply]
Glad to hear that!

**Reply by FabioEight** [reply]
Hello, I have the same issue also while connecting locally (direct IP connection). I had to set a client secret to login successfully!

**Reply by jasonmhite** [reply]
Did anybody figure what reverse proxy configuration is needed for oauth? I can similarly get things to work connecting directly to the container, but connecting regularly like that is not an option for me so I need to figure out what is going wrong.

**Reply by victorbalssa** [reply]
Hey @jasonmhite, I would recommend using the Personal Token Authentication method rather than trying to configure the OAuth.
Tokens are stored on secure storage (keychains) if security is a concern.

**Reply by Shmood00** [reply]
Hi @victorbalssa ,

I'm unfortunately unable to connect using OAuth or personal token. I receive the exact same error @kjakobczyk experienced using both options.

Any idea as to what may be causing this? I don't have Firefly setup behind a proxy, I'm connecting directly.

**Reply by victorbalssa** [reply]
Hi @Shmood00,

A couple of things you can try:
1. Try OAuth with a Client Secret ("Confidential" checked when creating a new client) and don't forget to copy paste the Client Secret. 
2. Double check the new redirect URL, it should be: `abacusfiiiapp://redirect` 
3. Double check the Client ID and URL of fireflyIII
4. If still not working, try with only a "Personal Access Tokens" ([more info](https://github.com/victorbalssa/abacus/blob/updates/.github/HELP.md#personal-access-token))
5. Last chance, try without http(s) or directly `http://` (but I don't recommend this if your instance is public for security reasons).

**Reply by Shmood00** [reply]
Hi @victorbalssa ,

I've actually tried / done everything you have suggested and it still does not work.

**Reply by Shmood00** [reply]
Small update:

I removed my self signed certificate (don't worry, instance isn't public facing) and then OAuth with the secret worked. 

I guess it didn't like the self signed certificates?

**Reply by victorbalssa** [reply]
Yes, I don't know why but self signed certificates does not work atm.

**Reply by Shmood00** [reply]
Good to know, thanks for the assistance!

**Reply by ragavpr** [reply]
> Yes, I don't know why but self signed certificates does not work atm.

I've got the same error, I have my own Certificate Authority, added to android user CAs trust store.

I think something like allowing custom certificate chains manually may work, just like what Waterfly III does, as it also has its own upstream problems with Flutter, it asks for a Custom Certificate to trust (I just put the Root PEM and it worked there)

It seems, this issue is also tracked in the upstream, will keep an eye out.

**Reply by nick4333** [reply]
Is there any plan to address the issue with the reverse proxy, I am also seeing it using nginx as the proxy and iOS as the client. I have multiple services running, which is why I have the proxy in the first place. And just opening a port to address this is a bad long term solution.

I would prefer to have all my traffic go through the proxy so that I can use my tls certificates for security which are not self signed and going through the openSSL CA.

**Reply by victorbalssa** [reply]
Hey @nick4333,
If you have a self signed certificate installed it won't work atm.
I have a task on this one #246 to track my progress.

**Reply by nick4333** [reply]
It's not self signed. It's signed by OpenSSL, which is an open source trusted CA that allows you to get a signed cert for free.

**Reply by victorbalssa** [reply]
@nick4333 
Personal Access Token should work

**Reply by nick4333** [reply]
It doesn't, it's related to the reverse proxy as connecting directly by IP, as others have noted, works correctly with no issues. I have tried multiple times with a personal access token, disabling the tls certificates, and the only way I could get it to work was unsecure http, bypassing the reverse proxy.

My config isn't doing anything crazy either, just setting a forwarded for ip header, then sending the rest of the request to the application.

**Reply by victorbalssa** [reply]
@nick4333 I'm not sure how to debug this, a Personal Access Token is set in the `Authorization` header, if you forward it correctly and don't change the payload it should work normally 🤔

**Reply by danielriddell21** [reply]
@victorbalssa was there ever a resolution to @nick4333 ’s issue.

I seem to be experiencing the same thing

**Comment by bbgth** [comment]
@victorbalssa I am having trouble connecting to my firefly (docker) server via nginx reverse proxy. It was working in v0.11.2 (OTA) but when v0.12.4 was released, I get "Something went wrong. Failed to get accessToken, Network request failed"

I am running iOS 17 on 2 devices, and iOS 16 on another device. All having the same issues

I am able to connect if I use direct IP:Port and bypass nginx.

No SSL usage. HTTP only since it is running locally

No changes to nginx config since v0.11.2

nginx logs:

> "GET /oauth/authorize?code_challenge=<redacted>&code_challenge_method=<redacted>&client_secret=<redacted>&redirect_uri=abacusfiiiapp%3A%2F%2Fredirect&client_id=<redacted>&response_type=code&state=<redacted>HTTP/1.1" 302 5 "-" "<redacted>"

That's it.

Nginx config:
>server {
        listen 80;
        server_name <redacted>;
        location / {
                proxy_pass http://<redacted. docker name>:8080/;
                proxy_redirect off;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection "upgrade";
                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-Proto $scheme;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Host $host;
                proxy_set_header X-Forwarded-Server $host;
        }
}

I have also tried "Personal Access Token". It didn't even seem to hit the server. No log entry whatsoever.

Thoughts?

**Reply by danielriddell21** [reply]
I have the exact same issue. I am behind Authentik using Traefik. 

I am using https, but using a PAT doesnt seem to hit the server at all

I have made sure and can confirm the api paths are bypassed. Its almost like the app isnt sending a request when using PAT

**Reply by victorbalssa** [reply]
I will need some time to fix this we can follow it here #256

**Reply by bbgth** [reply]
@victorbalssa for me, PAT doesn’t hit server and normal OAUTH using client ID and client Secret (my previous v0.11.2) method no longer works with "Something went wrong. Failed to get accessToken, Network request failed" error

Just want to make sure this is captured too. #256 seems to be only for PAT issue

**Reply by bbgth** [reply]
@victorbalssa I’ve done a bit more testing. I noticed if I access it directly via IP, it works (http://123.456.789.1:1234)

port 1234 is directly pointed to the firefly docker container and doesn’t go through nginx reverse proxy

http://firefly.domain.tld:1234 where firefly.domain.tld is resolved to IP 123.456.789.1 doesn’t work despite it doesn’t go through nginx reverse proxy as well. 

http://firefly.domain.tld where firefly.domain.tld is resolved to IP 123.456.789.1 doesn’t work. This url goes through nginx reverse proxy which was working in a prior Abacus version.

Note: this is all on http (no ssl) since it is all only locally accessible

Note 2: firefly latest docker image

Note 3: this affects BOTH PAT AND OAUTH

Note 4: IP address is a local lan IP. Not an internet routable IP. DNS is resolved using local dns resolver (with custom dns names)

**Reply by paulcalabro** [reply]
I actually did pretty much the same test @bbgth did to troubleshoot #256 and arrived at the same results myself.
I noticed I saw no traffic via Burp Suite when using the "Log In" button, but saw traffic when clicking the "Create a new PAT" link 
above, which is opened independently in a browser. I too am use local LAN IP with a local DNS  resolver with custom DNS resource records. When I swapped out the hostname for an IP address and used a NC listener, I saw traffic.
